### ISSUE TITLE
pkg state module does not allow specifying variants for macports packages

### ISSUE THREAD

**Issue by aphor** [saltstack/salt#45967]
### Description of Issue/Question
See #45413 and PR #45462
#### example state
    hdf5:
      - pkg.installed:
        - variant: +openmpi-cxx-hl
        - refresh_db: false

### Steps to Reproduce Issue
The variant in the above example is non-default, and is ignored by the macport "port" command issued by the underlying call to the salt/modules/mac_ports.py module. As a result, macports installs the package with (default) wrong features and dependencies.

### Versions Report

    Salt Version:
               Salt: 2018.2.0-133-gb100865
     
    Dependency Versions:
               cffi: Not Installed
           cherrypy: Not Installed
           dateutil: Not Installed
          docker-py: Not Installed
              gitdb: Not Installed
          gitpython: Not Installed
              ioflo: Not Installed
             Jinja2: 2.10
            libgit2: Not Installed
            libnacl: Not Installed
           M2Crypto: Not Installed
               Mako: Not Installed
       msgpack-pure: Not Installed
     msgpack-python: 0.5.4
       mysql-python: Not Installed
          pycparser: Not Installed
           pycrypto: Not Installed
       pycryptodome: 3.4.12
             pygit2: Not Installed
             Python: 3.6.4 (default, Dec 21 2017, 20:33:17)
       python-gnupg: Not Installed
             PyYAML: 3.12
              PyZMQ: 17.0.0
               RAET: Not Installed
              smmap: Not Installed
            timelib: Not Installed
            Tornado: 4.5.3
                ZMQ: 4.1.6
     
    System Versions:
               dist:   
             locale: UTF-8
            machine: x86_64
            release: 16.7.0
             system: Darwin
            version: 10.12.6 x86_64

**Comment by aphor** [comment]
@KaiSzuttor if you have any additional difficulties with variants and pkg.installed states on macports systems, please add comments here.

**Comment by aphor** [comment]
I'm not sure, but I think this may require significant changes to salt/states/pkg.py or salt/modules/mac_ports.py in order to handle the variants' spec separate from the version string. It would be most pythonic to handle the package data from state declaration and repo/installed package lists as a dict per package eg. `{'name': 'hdf5', 'variant': {'openmpi'}, variant-rejected: {'cxx-hl'}}` but some extra logic will be required to properly compare the state intent to the available packages and the installed packages.

There's also some ambiguity: should the pkg.installed state contain variants which are different from the currently installed package, do we need to introspect the package info to run the port install command with "-variant" where there are variant conflicts with the variants specified in the pkg.installed state.sls? I think it would be easier in a first pass to NOT do this, and just handle the command error with a good exception and error log message. The given example from @KaiSzuttor seems to need this behavior though.

**Comment by gtmanfred** [comment]
I would say instead of providing the variant in a seperate argument, you could handle the variant name in the mac_ports pkg module?

Just look for the `+` sign, and split it to be pkgname and variant.

**Comment by aphor** [comment]
Variants in macports are more like version data than separate packages (with separate names). They get treated like macros for configure options at build time, mostly.

I don't think it would be less messy to clobber package names by appending variants because it's still a pain to handle cases where a minion has foo+sparkles installed and active, but a pkg.installed state asks for foo+sprinkles. If the sparkles conflicts with sprinkles, salt/states/pkg.py should treat the request for foo+sprinkles as foo-sparkles+sprinkles. If there is no conflict, it should attempt to install foo+sparkles+sprinkles.

This is me expressing my inner Steve Jobs about how this *should* work, but getting this right seems like it might be super important to at least a few people, maybe more.

**Comment by aphor** [comment]
There is demand besides macports for refactoring the list_packages data at the heart.. #45948

I think this issue is blocked on the results of #45948. At this point, I think I am in favor of it.

**Comment by KaiSzuttor** [comment]
thank you @aphor for the new issue. I think what is in some way special to macports is that you foo+sparkles and foo+sprinkles can be installed at the same time. Macports resolves this by symbolic links (```port select --set ...```) which is not yet covered by the states module.
